File: quan.cs/Test/Config/Item/ItemConfig.cs
using Newtonsoft.Json.Linq;
using Quan.Utils;
using Quan.Config;
using System;
using System.Collections.Generic;
using System.Collections.Immutable;

namespace Test.Config.Item
{
    /// <summary>
	/// 道具<br/>
	/// 代码自动生成，请勿手动修改
	/// </summary>
    public class ItemConfig : ConfigBase
    {
        /// <summary>
        /// ID
        /// </summary>
        public readonly int id;

        /// <summary>
        /// 常量Key
        /// </summary>
        public readonly string key;

        /// <summary>
        /// 名字
        /// </summary>
        public readonly string name;

        /// <summary>
        /// 类型
        /// </summary>
        public readonly ItemType type;

        /// <summary>
        /// 使用效果
        /// </summary>
        public readonly UseEffect useEffect;

        /// <summary>
        /// 奖励
        /// </summary>
        public readonly Reward reward;

        /// <summary>
        /// List
        /// </summary>
        public readonly IList<int> list;

        /// <summary>
        /// Set
        /// </summary>
        public readonly ISet<int> set;

        /// <summary>
        /// Map
        /// </summary>
        public readonly IDictionary<int, int> map;

        /// <summary>
        /// 生效时间
        /// </summary>
        public readonly DateTime effectiveTime;

        public readonly string effectiveTime_;


        public ItemConfig(JObject json) : base(json)
        {
            id = json["id"]?.Value<int>() ?? default;
            key = json["key"]?.Value<string>() ?? "";
            name = json["name"]?.Value<string>() ?? "";
            type = (ItemType) (json["type"]?.Value<int>() ?? default);
            useEffect = json.ContainsKey("useEffect") ? UseEffect.Create(json["useEffect"].Value<JObject>()) : null;
            reward = json.ContainsKey("reward") ? Reward.Create(json["reward"].Value<JObject>()) : null;

            var list1 = json["list"]?.Value<JArray>();
            var list2 = ImmutableList<int>.Empty;
            if (list1 != null)
            {
                foreach (var listValue in list1)
                {
                    list2 =list2.Add(listValue.Value<int>());
                }
            }
            list = list2;

            var set1 = json["set"]?.Value<JArray>();
            var set2 = ImmutableHashSet<int>.Empty;
            if (set1 != null)
            {
                foreach (var setValue in set1)
                {
                    set2 =set2.Add(setValue.Value<int>());
                }
            }
            set = set2;

            var map1 = json["map"]?.Value<JObject>();
            var map2 = ImmutableDictionary<int, int>.Empty;
            if (map1 != null)
            {
                foreach (var mapKeyValue in map1)
                {
                    map2 = map2.Add(int.Parse(mapKeyValue.Key), mapKeyValue.Value.Value<int>());
                }
            }
            map = map2;

            effectiveTime = ToDateTime(json["effectiveTime"]?.Value<long>() ?? default);
            effectiveTime_ = json["effectiveTime_"]?.Value<string>() ?? "";
        }

        protected override ConfigBase Create(JObject json)
        {
            return new ItemConfig(json);
        }

        public override string ToString()
        {
            return "ItemConfig{" +
                   "id=" + id.ToString2() +
                   ",key='" + key + '\'' +
                   ",name='" + name + '\'' +
                   ",type=" + type.ToString2() +
                   ",useEffect=" + useEffect.ToString2() +
                   ",reward=" + reward.ToString2() +
                   ",list=" + list.ToString2() +
                   ",set=" + set.ToString2() +
                   ",map=" + map.ToString2() +
                   ",effectiveTime='" + effectiveTime_ + '\'' +
                   '}';
        }

        // 所有ItemConfig
        private static volatile IList<ItemConfig> _configs = new List<ItemConfig>();

        // 索引:ID
        private static volatile IDictionary<int, ItemConfig> _idConfigs = new Dictionary<int, ItemConfig>();

        // 索引:常量Key
        private static volatile IDictionary<string, ItemConfig> _keyConfigs = new Dictionary<string, ItemConfig>();

        // 索引:类型
        private static volatile IDictionary<ItemType, IList<ItemConfig>> _typeConfigs = new Dictionary<ItemType, IList<ItemConfig>>();

        public static IList<ItemConfig> GetAll()
        {
            return _configs;
        }

        public static IDictionary<int, ItemConfig> GetIdAll()
        {
            return _idConfigs;
        }

        public static ItemConfig Get(int id)
        {
            _idConfigs.TryGetValue(id, out var result);
            return result;
        }

        public static IDictionary<string, ItemConfig> GetKeyAll()
        {
            return _keyConfigs;
        }

        public static ItemConfig GetByKey(string key)
        {
            _keyConfigs.TryGetValue(key, out var result);
            return result;
        }

        public static IDictionary<ItemType, IList<ItemConfig>> GetTypeAll()
        {
            return _typeConfigs;
        }

        public static IList<ItemConfig> GetByType(ItemType type)
        {
            _typeConfigs.TryGetValue(type, out var result);
            return result ?? ImmutableList<ItemConfig>.Empty;
        }


        public static void Load(IList<ItemConfig> configs)
        {
            IDictionary<int, ItemConfig> idConfigs = new Dictionary<int, ItemConfig>();
            IDictionary<string, ItemConfig> keyConfigs = new Dictionary<string, ItemConfig>();
            IDictionary<ItemType, IList<ItemConfig>> typeConfigs = new Dictionary<ItemType, IList<ItemConfig>>();

            foreach (var config in configs)
            {
                Load(idConfigs, config, config.id);
                Load(keyConfigs, config, config.key);
                Load(typeConfigs, config, config.type);
            }

            configs = configs.ToImmutableList();
            idConfigs = ToImmutableDictionary(idConfigs);
            keyConfigs = ToImmutableDictionary(keyConfigs);
            typeConfigs = ToImmutableDictionary(typeConfigs);

            _configs = configs;
            _idConfigs = idConfigs;
            _keyConfigs = keyConfigs;
            _typeConfigs = typeConfigs;
        }
    }
}

File: quan.java/quan-config/src/test/java/quan/config/item/ItemConfig.java
package quan.config.item;

import com.alibaba.fastjson.*;
import quan.config.*;
import quan.config.loader.ConfigLoader;
import java.util.*;

/**
 * 道具<br/>
 * 代码自动生成，请勿手动修改
 */
public class ItemConfig extends Config {

    /**
     * ID
     */
    public final int id;

    /**
     * 常量Key
     */
    public final String key;

    /**
     * 名字
     */
    public final String name;

    /**
     * 类型
     */
    public final ItemType type;

    /**
     * 使用效果
     */
    public final UseEffect useEffect;

    /**
     * 奖励
     */
    public final Reward reward;

    /**
     * List
     */
    public final List<Integer> list;

    /**
     * Set
     */
    public final Set<Integer> set;

    /**
     * Map
     */
    public final Map<Integer, Integer> map;

    /**
     * 生效时间
     */
    public final Date effectiveTime;

    /**
     * 生效时间
     */
    public final String effectiveTime_;


    public ItemConfig(JSONObject json) {
        super(json);

        this.id = json.getIntValue("id");
        this.key = json.getOrDefault("key", "").toString();
        this.name = json.getOrDefault("name", "").toString();
        this.type = ItemType.valueOf(json.getIntValue("type"));

        JSONObject useEffect = json.getJSONObject("useEffect");
        if (useEffect != null) {
            this.useEffect = UseEffect.create(useEffect);
        } else {
            this.useEffect = null;
        }

        JSONObject reward = json.getJSONObject("reward");
        if (reward != null) {
            this.reward = Reward.create(reward);
        } else {
            this.reward = null;
        }

        JSONArray list$1 = json.getJSONArray("list");
        List<Integer> list$2 = new ArrayList<>();
        if (list$1 != null) {
            for (int i = 0; i < list$1.size(); i++) {
                list$2.add(list$1.getInteger(i));
            }
        }
        this.list = Collections.unmodifiableList(list$2);

        JSONArray set$1 = json.getJSONArray("set");
        Set<Integer> set$2 = new HashSet<>();
        if (set$1 != null) {
            for (int i = 0; i < set$1.size(); i++) {
                set$2.add(set$1.getInteger(i));
            }
        }
        this.set = Collections.unmodifiableSet(set$2);

        JSONObject map$1 = json.getJSONObject("map");
        Map<Integer, Integer> map$2 = new HashMap<>();
        if (map$1 != null) {
            for (String map$Key : map$1.keySet()) {
                map$2.put(Integer.valueOf(map$Key), map$1.getInteger(map$Key));
            }
        }
        this.map = Collections.unmodifiableMap(map$2);

        this.effectiveTime = json.getDate("effectiveTime");
        this.effectiveTime_ = json.getOrDefault("effectiveTime_", "").toString();
    }

    @Override
    public ItemConfig create(JSONObject json) {
        return new ItemConfig(json);
    }

    @Override
    public String toString() {
        return "ItemConfig{" +
                "id=" + id +
                ",key='" + key + '\'' +
                ",name='" + name + '\'' +
                ",type=" + type +
                ",useEffect=" + useEffect +
                ",reward=" + reward +
                ",list=" + list +
                ",set=" + set +
                ",map=" + map +
                ",effectiveTime='" + effectiveTime_ + '\'' +
                '}';

    }


    //所有ItemConfig
    private static volatile List<ItemConfig> _configs = Collections.emptyList();

    //索引:ID
    private static volatile Map<Integer, ItemConfig> _idConfigs = Collections.emptyMap();

    //索引:常量Key
    private static volatile Map<String, ItemConfig> _keyConfigs = Collections.emptyMap();

    //索引:类型
    private static volatile Map<ItemType, List<ItemConfig>> _typeConfigs = Collections.emptyMap();

    public static List<ItemConfig> getAll() {
        return _configs;
    }

    public static Map<Integer, ItemConfig> getIdAll() {
        return _idConfigs;
    }

    public static ItemConfig get(int id) {
        return _idConfigs.get(id);
    }

    public static Map<String, ItemConfig> getKeyAll() {
        return _keyConfigs;
    }

    public static ItemConfig getByKey(String key) {
        return _keyConfigs.get(key);
    }

    public static Map<ItemType, List<ItemConfig>> getTypeAll() {
        return _typeConfigs;
    }

    public static List<ItemConfig> getByType(ItemType type) {
        return _typeConfigs.getOrDefault(type, Collections.emptyList());
    }


    @SuppressWarnings({"unchecked"})
    private static List<String> load(List<ItemConfig> configs) {
        Map<Integer, ItemConfig> idConfigs = new HashMap<>();
        Map<String, ItemConfig> keyConfigs = new HashMap<>();
        Map<ItemType, List<ItemConfig>> typeConfigs = new HashMap<>();

        List<String> errors = new ArrayList<>();

        for (ItemConfig config : configs) {
            load(idConfigs, errors, config, true, Collections.singletonList("id"), config.id);
            if (!config.key.isEmpty()) {
                load(keyConfigs, errors, config, true, Collections.singletonList("key"), config.key);
            }
            load(typeConfigs, errors, config, false, Collections.singletonList("type"), config.type);
        }

        configs = Collections.unmodifiableList(configs);
        idConfigs = unmodifiableMap(idConfigs);
        keyConfigs = unmodifiableMap(keyConfigs);
        typeConfigs = unmodifiableMap(typeConfigs);

        ItemConfig._configs = configs;
        ItemConfig._idConfigs = idConfigs;
        ItemConfig._keyConfigs = keyConfigs;
        ItemConfig._typeConfigs = typeConfigs;

        return errors;
    }

    static {
        ConfigLoader.registerLoadFunction(ItemConfig.class, ItemConfig::load);
    }

}
